logsmanagement/parser.c

```diff
@@ -0,0 +1,1622 @@
+                else if (unlikely(regex_rc == REG_NOMATCH)) {
+                    #if ENABLE_PARSE_WEB_LOG_LINE_DEBUG
+                    error("SSL_CIPHER_SUITE is invalid");
+                    #endif
+                    log_line_parsed->ssl_cipher[0] = '\0';
+                    log_line_parsed->parsing_errors++;
+                }
+                else {
+                    size_t err_msg_size = regerror(regex_rc, &cipher_suite_regex, NULL, 0);
+                    char *err_msg = mallocz(err_msg_size);
+                    regerror(regex_rc, &cipher_suite_regex, err_msg, err_msg_size);
+                    error("cipher_suite_regex error:%s", err_msg);
+                    freez(err_msg);
+                    m_assert(0, "cipher_suite_regex has failed");
+                }
+            }
+
+            #if ENABLE_PARSE_WEB_LOG_LINE_DEBUG
+            debug(D_LOGS_MANAG, "Extracted SSL_CIPHER_SUITE:%s", log_line_parsed->ssl_cipher);
+            #endif
+
+            goto next_item;
+        }
+
+        if(fields_format[i] == TIME){
+
+#if SKIP_WEB_LOG_TIME_PARSING
+            while(*offset != ']') {offset++;};
+            i++;
+            offset++;
+            goto next_item;
+#endif // SKIP_WEB_LOG_TIME_PARSING
+
+            #if ENABLE_PARSE_WEB_LOG_LINE_DEBUG
+            debug(D_LOGS_MANAG, "Item %d (type: TIME - 1st of 2 fields):%.*s", i, (int)field_size, field);
+            #endif
+
+            // TODO: What if TIME equals = ??
+            // if(field[0] ==  '-' && field_size == 1) { 
+            //     log_line_parsed->timestamp = 0;
+            //     log_line_parsed->parsing_errors++;
+            //     ++i;
+            //     goto next_item;
+            // }
+
+            char *datetime = field;
+
+            if(memchr(datetime, '[', field_size)) {
+                datetime++;
+                field_size--;
+            }
+
+            struct tm ltm = {0};
+            char *second_part = strptime(datetime, "%d/%b/%Y:%H:%M:%S", &ltm);
+            if(unlikely(second_part == NULL)){
+                printf("TIME datetime parsing failed\n");
+                log_line_parsed->timestamp = 0;
+                log_line_parsed->parsing_errors++;
+                // ++i;
+                goto next_item;
+            }
+
+            /* Deal with 2nd part of datetime i.e. timezone
+             * TODO: Error handling in case the second part is not timezone?? */
+
+            m_assert(*second_part == ' ', "Invalid TIME second_part");
+            m_assert(*++second_part == '+', "Invalid TIME second_part");
+
+            char *second_part_end = ++second_part;
+            while(*second_part_end != ']' && *second_part_end != ' ' && *second_part != '\n') {second_part_end++; };
+            size_t second_part_size = second_part_end - second_part;
+
+            // TODO: Get rid of strndup() - find alternative that doesn't require strtol()
+            second_part = strndup(second_part, second_part_size); 
```

Comment:
Why are not using `strdupz` here?

Reply:
I am not exactly sure but I think that `second_part` is not null-terminated, so I need to use `second_part_size`. We don't have an `strndupz()` function, so I am using `strndup` (however, it requires `#define _XOPEN_SOURCE 700`, so it might not be ideal).

Reply:
All right, I am marking this as resolved.